applications/WindEngineeringApplication/custom_utilities/model_subdivision_utilities.h

```diff
@@ -0,0 +1,195 @@
+     *  in rSubModelParts. If sortFunction returns NULL, the node is skipped and is not added to any of the model parts. 
+     */
+    static void SortNodes(
+        ModelPart& rModelPart,
+        std::function<ModelPart*(const Node<3>&)> sortFunction,
+        std::vector<ModelPart*>& rSubModelParts);
+
+    /** Sort elements into sub model parts
+     *  @param rModelPart model part containing elements to sort
+     *  @param sortFunction function that decides which sub model part a element is added to (NULL means it is skipped)
+     *  @param rSubModelParts set of sub model parts to sort elements into
+     *  @details calls sortFunction for every element in rModelPart, and adds each of them to a sub model part
+     *  in rSubModelParts. If sortFunction returns NULL, the element is skipped and is not added to any of the model parts. 
+     */
+    static void SortElements(
+        ModelPart& rModelPart,
+        std::function<ModelPart*(const Element&)> sortFunction,
+        std::vector<ModelPart*>& rSubModelParts);
+
+    /** Sort conditions into sub model parts
+     *  @param rModelPart model part containing conditions to sort
+     *  @param sortFunction function that decides which sub model part a condition is added to (NULL means it is skipped)
+     *  @param rSubModelParts set of sub model parts to sort conditions into
+     *  @details calls sortFunction for every condition in rModelPart, and adds each of them to a sub model part
+     *  in rSubModelParts. If sortFunction returns NULL, the condition is skipped and is not added to any of the model parts. 
+     */
+    static void SortConditions(
+        ModelPart& rModelPart,
+        std::function<ModelPart*(const Condition&)> sortFunction,
+        std::vector<ModelPart*>& rSubModelParts);
+
+protected:
+    /** A set of integer vectors that can be pushed to in parallel
+     *  @details each vector is assigned to a sub model part and has their own mutex.
+     */
+    class ThreadSafeIndexSet
+    {
+    public:
+        using IndexType = ModelPart::IndexType;
+
+        ThreadSafeIndexSet(std::vector<ModelPart*>& rSubModelParts);
+
+        /** Push to a vector that is assigned to a specific model part
+         *  @param pSubModelPart pointer to the desired model part (NULL is valid too)
+         *  @param value integer to be pushed
+         *  @note pSubModelPart==NULL is valid and means value will not be pushed anywhere
+         */
+        void Push(ModelPart* pSubModelPart, IndexType value);
+
+        /// Only to be used with ModelPart::AddNodes, ModelPart::AddElements, and ModelPart::AddConditions
+        template <class TFunction = std::function<void(ModelPart*, const std::vector<IndexType>&, IndexType)>>
+        void Apply(TFunction function);
+
+    private:
+        std::map<ModelPart*,
+                 std::pair<std::vector<IndexType>,
+                           unique_ptr<std::mutex>>> mIndexSets;
```

Comment:
```suggestion
                           unique_ptr<LockObject>>> mIndexSets;
```

this should use our `LockObject`. Reason: OpenMP does not guarantee that it is compatible! Some implementations afaik use the same OS-functions to implemente both `mutex`  and `omp_lock`, but it is not sure. And those errors would be very nasty to debug

The `LockObject` uses the correct mechanism under the hood (see e.g. #8161 which I still haven't finished)
This way you are sure the correct lock is used